File: my_vpn/app/src/main/java/com/myvpn/simple/ui/AppExclusionAdapter.java
package com.myvpn.simple.ui;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CheckBox;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.myvpn.simple.R;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class AppExclusionAdapter extends RecyclerView.Adapter<AppExclusionAdapter.ViewHolder> {
    
    public interface OnAppExclusionChangeListener {
        void onAppExclusionChanged(String packageName, boolean isExcluded);
    }
    
    private List<AppInfo> appList = new ArrayList<>();
    private List<AppInfo> filteredAppList = new ArrayList<>();
    private OnAppExclusionChangeListener listener;
    private String searchQuery = "";
    
    public void setOnAppExclusionChangeListener(OnAppExclusionChangeListener listener) {
        this.listener = listener;
    }
    
    public void setAppList(List<AppInfo> appList) {
        this.appList = new ArrayList<>(appList);
        Collections.sort(this.appList, (a1, a2) -> {
            // 1. 先按排除状态排序 (选中的在前面)
            if (a1.isExcluded() != a2.isExcluded()) {
                return a1.isExcluded() ? -1 : 1;
            }
            // 2. 再按应用类型排序 (用户应用在前面)
            if (a1.isSystemApp() != a2.isSystemApp()) {
                return a1.isSystemApp() ? 1 : -1;
            }
            // 3. 最后按应用名排序
            return a1.getAppName().compareToIgnoreCase(a2.getAppName());
        });
        applyFilter();
    }
    
    public void setSearchQuery(String query) {
        this.searchQuery = query.toLowerCase();
        applyFilter();
    }
    
    private void applyFilter() {
        filteredAppList.clear();
        for (AppInfo app : appList) {
            if (searchQuery.isEmpty() || 
                app.getAppName().toLowerCase().contains(searchQuery) ||
                app.getPackageName().toLowerCase().contains(searchQuery)) {
                filteredAppList.add(app);
            }
        }
        notifyDataSetChanged();
    }
    
    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext())
            .inflate(R.layout.item_app_exclusion, parent, false);
        return new ViewHolder(view);
    }
    
    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        AppInfo app = filteredAppList.get(position);
        
        holder.appIcon.setImageDrawable(app.getIcon());
        holder.appName.setText(app.getAppName());
        holder.packageName.setText(app.getPackageName());
        
        // 显示应用类型标签
        if (app.isSystemApp()) {
            holder.appType.setText("系统");
            holder.appType.setVisibility(View.VISIBLE);
        } else {
            holder.appType.setVisibility(View.GONE);
        }
        
        holder.checkBox.setOnCheckedChangeListener(null);
        holder.checkBox.setChecked(app.isExcluded());
        
        holder.checkBox.setOnCheckedChangeListener((buttonView, isChecked) -> {
            app.setExcluded(isChecked);
            if (listener != null) {
                listener.onAppExclusionChanged(app.getPackageName(), isChecked);
            }
            // 重新排序
            setAppList(appList);
        });
        
        holder.itemView.setOnClickListener(v -> {
            holder.checkBox.setChecked(!holder.checkBox.isChecked());
        });
    }
    
    @Override
    public int getItemCount() {
        return filteredAppList.size();
    }
    
    public static class ViewHolder extends RecyclerView.ViewHolder {
        ImageView appIcon;
        TextView appName;
        TextView packageName;
        TextView appType;
        CheckBox checkBox;
        
        public ViewHolder(@NonNull View itemView) {
            super(itemView);
            appIcon = itemView.findViewById(R.id.app_icon);
            appName = itemView.findViewById(R.id.app_name);
            packageName = itemView.findViewById(R.id.package_name);
            appType = itemView.findViewById(R.id.app_type);
            checkBox = itemView.findViewById(R.id.checkbox);
        }
    }
}

File: kt/app/src/main/kotlin/com/myvpn/simple/ui/AppExclusionAdapter.kt
package com.myvpn.simple.ui

import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.CheckBox
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.myvpn.simple.R

class AppExclusionAdapter : RecyclerView.Adapter<AppExclusionAdapter.ViewHolder>() {

    companion object {
        private const val TAG = "AppExclusionAdapter"
    }

    fun interface OnAppExclusionChangeListener {
        fun onAppExclusionChanged(packageName: String, isExcluded: Boolean)
    }

    private var appList: MutableList<AppInfo> = mutableListOf()
    private var filteredAppList: MutableList<AppInfo> = mutableListOf()
    private var listener: OnAppExclusionChangeListener? = null
    private var searchQuery: String = ""

    fun setOnAppExclusionChangeListener(listener: OnAppExclusionChangeListener) {
        this.listener = listener
    }

    fun setAppList(appList: List<AppInfo>) {
        this.appList = appList.toMutableList().apply {
            sortWith(compareBy<AppInfo> { !it.isExcluded }  // 选中的在前面
                .thenBy { it.isSystemApp }  // 用户应用在前面
                .thenBy { it.appName.lowercase() })  // 按应用名排序
        }
        applyFilter()
    }

    fun setSearchQuery(query: String) {
        val newQuery = query.lowercase().trim()
        Log.d(TAG, "setSearchQuery called: '$newQuery' (was: '$searchQuery')")
        if (searchQuery != newQuery) {
            searchQuery = newQuery
            applyFilter()
        }
    }

    private fun applyFilter() {
        val previousSize = filteredAppList.size
        filteredAppList = if (searchQuery.isEmpty()) {
            appList.toMutableList()
        } else {
            appList.filter { app ->
                app.appName.lowercase().contains(searchQuery) ||
                        app.packageName.lowercase().contains(searchQuery)
            }.toMutableList()
        }

        Log.d(TAG, "applyFilter: query='$searchQuery', total=${appList.size}, filtered=${filteredAppList.size}")

        // 使用更精确的通知方法
        if (previousSize == 0 && filteredAppList.isNotEmpty()) {
            notifyItemRangeInserted(0, filteredAppList.size)
        } else if (filteredAppList.isEmpty() && previousSize > 0) {
            notifyItemRangeRemoved(0, previousSize)
        } else {
            notifyDataSetChanged()
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view = LayoutInflater.from(parent.context)
            .inflate(R.layout.item_app_exclusion, parent, false)
        return ViewHolder(view)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val app = filteredAppList[position]

        holder.apply {
            appIcon.setImageDrawable(app.icon)
            appName.text = app.appName
            packageName.text = app.packageName

            // 显示应用类型标签
            if (app.isSystemApp) {
                appType.text = "系统"
                appType.visibility = View.VISIBLE
            } else {
                appType.visibility = View.GONE
            }

            checkBox.setOnCheckedChangeListener(null)
            checkBox.isChecked = app.isExcluded

            checkBox.setOnCheckedChangeListener { _, isChecked ->
                app.isExcluded = isChecked
                listener?.onAppExclusionChanged(app.packageName, isChecked)
                // 更新appList中的对应项
                val originalApp = appList.find { it.packageName == app.packageName }
                originalApp?.isExcluded = isChecked
                // 重新排序
                setAppList(appList)
            }

            itemView.setOnClickListener {
                checkBox.isChecked = !checkBox.isChecked
            }
        }
    }

    override fun getItemCount(): Int = filteredAppList.size

    class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val appIcon: ImageView = itemView.findViewById(R.id.app_icon)
        val appName: TextView = itemView.findViewById(R.id.app_name)
        val packageName: TextView = itemView.findViewById(R.id.package_name)
        val appType: TextView = itemView.findViewById(R.id.app_type)
        val checkBox: CheckBox = itemView.findViewById(R.id.checkbox)
    }
}
